doc/TACACSPLUS_PASSKEY_ENCRYPTION.md

```diff
@@ -0,0 +1,87 @@
+The revised data handling procedure among the modules is outlined as follows:
+1. When a user configures the TACACS passkey using the SONIC CLI, it will now be securely stored in encrypted format instead of plaintext.
+2. Subsequently, the HostCfg Enforcer module retrieves this encrypted key. However, before writing it into the PAM configuration file(s), the hostCfgd module decrypts it.
+```
+       +-------+  +---------+
+       |  SSH  |  | Console |
+       +---+---+  +----+----+
+           |           |   
++----------v-----------v---------+                                      +---------------------+
+| AUTHENTICATION                 |                                      |                     |
+|   +-------------------------+  |  Decrypted passkey                   |  +------------+     |
+|   | PAM Configuration Files <-----------------------------------------+  | AAA Config |     |
+|   +-------------------------+  |                                      |  +------------+     |
+|                                |                                      |                     |
+|         +-------------+        |                                      |  HostCfg Enforcer   |
+|         |     PAM     |        |                                      +----------^----------+
+|         |  Libraries  |        |                                                 |
+|         +-------------+        |                                                 | Encrypted passkey
++---------------+----------------+                                                 |
+                |                                                                  |
+           +----v----+                                                     +-------+--------+
+           |         |             Encrypted passkey                       |                |
+           |   CLI   +------------------------------------------------------>    ConifgDB   |
+           |         |                                                     |                |
+           +---------+                                                     +----------------+
+```
+This decryption step is crucial because the login or SSH daemon references the PAM config file to verify the TACACS secret / passkey. If it remains encrypted, the SSH daemon will be unable to recognize the passkey, leading to login failures. The depicted block diagram clearly showcases the enhanced capabilities of the existing submodules.
+### Implementation details
+The implementation stands on four key pillars.
+1. OPENSSL toolkit is used for encryption / decryption.
+2. aes-128-cbc is the encoding format used for encryption / decryption.
+3. A unique device admin password (encrypted hash from linux shadow file) will be used as a salt/password to encrypt/decrypt the configured pass key in ConfigDB.
```

Comment:
How to handle password change?

There was a password hardening feature will make the password expired: https://github.com/sonic-net/SONiC/blob/master/doc/passw_hardening/hld_password_hardening.md?plain=1

Reply:
Yeah, the encryption and decryption is solely based on the "admin" password only. If the password expires and it has been changed, the network admin needs to generate a new encrypted passkey using the new admin password hash present in the shadow file.

Reply:
@nmoray , I think the encrypted key update should be automatically done when admin password changed.

If not, when the admin password changed, device can't connect with tacacs server anymore, which will break many automation service in production environment.

Reply:
@liuh-80 Automating this would not be straight forward as we need to modify "passwd" linux command to achieve the same.

We can document a requirement that this implementation is based on the admin password present in shadow file. Thus, if the operator is planing to change the "admin" password, he / she has to propagate that change to all the devices.

Reply:
So there will be added effort to adjust the hash key and perform the encryption-decryption once we change the admin password?

Reply:
@Yarden-Z only the encrypted key inside CONFIG_DB needs to be changed and it will be same across all the devices so operator simply needs to propagate it to all the devices.

Reply:
> @Yarden-Z only the encrypted key inside CONFIG_DB needs to be changed and it will be same across all the devices so operator simply needs to propagate it to all the devices.

I'm unsure if this is the only action required.
Let's say we have the tacplus secret encrypted with password "123456", which is the admin password.
If a user changes the admin password - we need to first decrypt the tacacs key and then re-encrypt it with the new admin password, otherwise - this data is lost.

Reply:
@Yarden-Z why do we need to decrypt the passkey? Here passkey is not changing, the encryption password is changing.

Reply:
Passkey in plaintext would be the same right!

Reply:
Yes, passkey in plaintext should remain the same. Encryption password is changing.
If we used this encryption password - then the ciphertext won't be recoverable.
Per the HLD: 
`When a user configures the TACACS passkey using the SONIC CLI, it will now be securely stored in encrypted format instead of plaintext.`

So, if we are storing it in ciphertext, and the encryption key is changed (admin password) then the ciphertext encryption needs to change as well.
Am I missing something here?

Reply:
okay the flow is like this,

When we encrypt a tacplus passkey with admin password (assume "1234"), it writes an encrypted hash in CONFIG_DB. Same time the encrypted key gets decrypted by hostcfgd and saved into "common-auth-sonic" file in plaintext format (here we don't need to do anything manually). Now let's say. operator needs to change the admin password to "4321" then he / she needs to write a new tacplus passkey in CONFIG_DB by simply running the "config tacacs passkey <str>" CLI. This way, a new encrypted key will get created in CONFIG_DB and the plaintext version of it will be stored in common-auth-sonic file. 

here, the operator only need to look after the encryption part. Decryption will happen automatically. I hope I am able to clarify. :-)

Reply:
IMO, every organization has different policy/requirements in terms of periodically changing the admin/user passwords of the devices, we should not depend on admin password for encryption key, rather follow the behavior (i.e have a separate master password configuration and default could be system MAC) of other popular NOS for SONiC.

References:
https://www.juniper.net/documentation/us/en/software/junos/user-access/topics/topic-map/master-password-configuration-encryption.html
https://www.cisco.com/c/en/us/support/docs/security-vpn/ipsec-negotiation-ike-protocols/46420-pre-sh-keys-ios-rtr-cfg.html

Reply:
I agree with the suggestion of having a "master password" or some global password for encrypting sensitive data.
This can later on be re-used for other sensitive items such as radius passkey or LDAP passkey (which is coming shortly).
In order to define this "master password" - we'll need to think of a way to create, store it and manage it if required.

In addition - if we would like to scale up this solution (add it other relevant areas in the system where we need to protect sensitive information) we do not want a user performing all those extra actions in the system (currently, it's 1 place, tomorrow it will be 5-10 different configs to change).

Reply:
@venkatmahalingam @Yarden-Z okay understood. I have referred the above link and decided to go with similar implementation where, I will introduce a 'key_encrypt' flag in tacplus table to enable this feature. If enabled, while adding the passkey, it will ask for a password (after entering "config tacacs passkey <str>" command) for the encryption. This password will not be visible on to the console (while entering). It will be stored in /etc/tacplus_pass file (similar to /etc/tacplus_user or /etc/tacplus_nss.conf) in the background and will be referred during the decryption process at hostcfgd.

Additionally, to make this implementation backward compatible, I will keep 'key_encrypt' flag as optional. Like, if it is absent or explicitly marked as False, there will not be any passkey encryption. Passkey will be stored in plaintext for mat only inside CONFIG_DB. 

I hope, now we can conclude on the HLD. :-)

Reply:
> If enabled, while adding the passkey, it will ask for a password (after entering "config tacacs passkey " command) for the encryption.

I think @Yarden-Z 's previous comment is asking for a more global solution, and should be one that is portable to other solutions like RADIUS, LDAP, and possibly other features that require symmetrically encrypted passkeys like NTP, OSPF/BGP, etc.

@Yarden-Z  also said:
> In order to define this "master password" - we'll need to think of a way to create, store it and manage it if required.

This might warrant another HLD!

Reply:
> @venkatmahalingam @Yarden-Z @liuh-80 okay understood. I have referred the above link and decided to go with similar implementation where, I will introduce a 'key_encrypt' flag in tacplus table to enable this feature. If enabled, while adding the passkey, it will ask for a password (after entering "config tacacs passkey " command) for the encryption. This password will not be visible on to the console (while entering). It will be stored in /etc/tacplus_pass file (similar to /etc/tacplus_user or /etc/tacplus_nss.conf) in the background and will be referred during the decryption process at hostcfgd.
> 
> Additionally, to make this implementation backward compatible, I will keep 'key_encrypt' flag as optional. Like, if it is absent or explicitly marked as False, there will not be any passkey encryption. Passkey will be stored in plaintext format only inside CONFIG_DB.
> 
> I hope, now we can conclude on the HLD. :-)

@venkatmahalingam what are your thoughts?

Reply:
> > If enabled, while adding the passkey, it will ask for a password (after entering "config tacacs passkey " command) for the encryption.
> 
> I think @Yarden-Z 's previous comment is asking for a more global solution, and should be one that is portable to other solutions like RADIUS, LDAP, and possibly other features that require symmetrically encrypted passkeys like NTP, OSPF/BGP, etc.
> 
> @Yarden-Z also said:
> 
> > In order to define this "master password" - we'll need to think of a way to create, store it and manage it if required.
> 
> This might warrant another HLD!

@jeff-yin Please let us know if you have any better way for encrypting the passkey? For having a same encryption password across all the features (REDIUS, TACACS, LDAP, etc), we can maintain a separate file at /etc/ to store the password similar to tacplus_nss.conf and same password will be used to encrypt the passkey. We will not ask for any password at the time of configuring any of these feature. All feature will by default read a file from /etc/ to encrypt / decrypt the passkey. Additionally, this encryption feature will be enabled / disabled with the help of a flag and this will be there in all the relevant  tables like, TACPLUS, REDIUS etc.

Reply:
This way, we can have MASTER password for the passkey encryption and it will not have any dependency on any other module.

Reply:
The best approach would be to introduce an infra that can provide encryption/decryption as a service across SONiC which will also take care of the key management and storage. This way, all protocols can use the service without worrying about management of the key. The service must be accessible to applications across dockers and host.

Reply:
> The best approach would be to introduce an infra that can provide encryption/decryption as a service across SONiC which will also take care of the key management and storage. This way, all protocols can use the service without worrying about management of the key. The service must be accessible to applications across dockers and host.

@nmoray This design looks generic and works for all protocols (TACACS+, RADIUS, LDAP, BGP..etc) that need the encryption/decryption. Refer below diagram for some details.

![image](https://github.com/sonic-net/SONiC/assets/34145258/5a10957a-e3a5-4074-9de4-8954f7f3fd58)

Reply:
@venkatmahalingam thank you, I have updated the HLD. Yes the design is generic, support all three protocols (TACACS+, RADIUS, LDAP, BGP..etc) that need the encryption/decryption. 202311 release is around the corner, could you please approve and merge the HLD, we will work on code PRs and get it done.